### PR TITLE
Change how we measure child views in UIViewBox

### PR DESCRIPTION
**Pull request by swankjesse**
Previously it broke when a view changed sizes after its initial measurement.

---
- [x] `CHANGELOG.md`'s "Unreleased" section has been updated, if applicable.

### CODE REVIEW

File: redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/UIViewBox.kt [hunk 1]
```diff
@@ -126,85 +127,95 @@ internal class UIViewBox :
...
-              itemHorizontalAlignment = childModifier.alignment
-            }
+      val frameWidth = frame.useContents { size.width }
+      val frameHeight = frame.useContents { size.height }
 
-            is VerticalAlignment -> {
-              itemVerticalAlignment = childModifier.alignment
-            }
+      var requestedWidth: CGFloat = Double.NaN
+      var requestedHeight: CGFloat = Double.NaN
```
Comment (swankjesse) [inline redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/UIViewBox.kt:145]: Switched from Double.MIN_VALUE (which is a possible actual value near zero) to Double.NaN which isn't a possible actual value.

File: redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/UIViewBox.kt [hunk 2]
```diff
@@ -126,85 +127,95 @@ internal class UIViewBox :
...
+        }
+      }
 
-          CrossAxisAlignment.Center -> x = (frame.useContents { size.width } - childWidth) / 2.0
+      // Initialize this to the view's width before any measurement. Use the frame's dimensions
+      // if the user didn't explicitly specify one. This is the final value for Stretch alignment.
+      var viewWidth = when {
+        !requestedWidth.isNaN() -> requestedWidth
+        else -> frameWidth
+      }
```
Comment (swankjesse) [inline redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/UIViewBox.kt:177]: Same as before, just `if` to `when` and renamed `childWidth` to `viewWidth`

File: redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/UIViewBox.kt [hunk 3]
```diff
@@ -126,85 +127,95 @@ internal class UIViewBox :
...
-            childHeight = frame.useContents { size.height }
-          }
+      // Measure the view if don't have an exact width or height.
+      val mustMeasureWidth = requestedWidth.isNaN() &&
+        itemHorizontalAlignment != CrossAxisAlignment.Stretch
+      val mustMeasureHeight = requestedHeight.isNaN() &&
+        itemVerticalAlignment != CrossAxisAlignment.Stretch
 
-          CrossAxisAlignment.Start -> y = 0.0
+      if (mustMeasureWidth || mustMeasureHeight) {
```
Comment (swankjesse) [inline redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/UIViewBox.kt:189]: This is the new part. Previously we were calling `view.sizeToFit()` at the top of this function to get a measurement as a side-effect; now we’re only doing measurement when we need to. And we’re asking for a measurement with appropriate parameters.

File: redwood-layout-uiview/RedwoodLayoutUIViewTests/__Snapshots__/UIViewBoxTestHost/testMargins.1.png
Comment (swankjesse) [inline redwood-layout-uiview/RedwoodLayoutUIViewTests/__Snapshots__/UIViewBoxTestHost/testMargins.1.png:1]: I broke this in the last PR and didn’t want to live with that! This is better.

**Review by JakeWharton — APPROVED** [review]

**Pull request merged**